### PR TITLE
[WIP] EEND with Clustering

### PR DESCRIPTION
**Pull request by Zenglinxiao**

### CODE REVIEW

File: eend/feature.py [hunk 1]
```diff
@@ -257,18 +268,32 @@ def get_labeledSTFT(
     T = np.zeros((Y.shape[0], n_speakers), dtype=np.int32)
 
     if use_speaker_id:
+        # NOTE all speaker contained in the kaldi data(train/eval set)
         all_speakers = sorted(kaldi_obj.spk2utt.keys())
         S = np.zeros((Y.shape[0], len(all_speakers)), dtype=np.int32)
 
+    if use_global_speaker_id:
+        global_speakers = kaldi_obj.global_speakers_with_silent
+        # Init with zeros, handle case: active speaker < S_local, 0 for silent
```
Comment (francoishernandez) [inline eend/feature.py:277]: Do they say they initialize with zeros in the paper?

File: eend/feature.py [hunk 2]
```diff
@@ -257,18 +268,32 @@ def get_labeledSTFT(
...
+        global_speaker_idx = np.zeros((n_speakers, ), dtype=np.int32)
+        # set active speaker in the chunk
+        # NOTE :n_speakers to remove speakers exceed n_speakers
+        _gsidx = [global_speakers.index(spk) for spk in speakers[:n_speakers]]
```
Comment (francoishernandez) [inline eend/feature.py:281]: No sure to understand this one.
speakers = speakers of only the recording?
Is this just a failsafe in case we have more speakers than expected by the model in a single recording?

File: eend/feature.py [hunk 3]
```diff
@@ -257,18 +268,32 @@ def get_labeledSTFT(
...
+        global_speaker_idx[:len(_gsidx)] = _gsidx
```
Comment (francoishernandez) [inline eend/feature.py:282]: I don't understand this one either.

File: eend/feature.py [hunk 4]
```diff
@@ -257,18 +268,32 @@ def get_labeledSTFT(
...
+
     for seg in filtered_segments:
+        # speaker_index of this segment
         speaker_index = speakers.index(kaldi_obj.utt2spk[seg['utt']])
```
Comment (francoishernandez) [inline eend/feature.py:286]: We build some global_speaker_idx above, but do not use it here. Normal?

File: eend/chainer_backend/infer.py [hunk 5]
```diff
@@ -81,33 +104,13 @@ def infer(args):
...
-                    shuffle=args.shuffle
-                )
-                if args.gpu >= 0:
-                    ys[0].to_cpu()
-                out_chunks.append(ys[0].data)
-                if args.save_attention_weight == 1:
-                    att_fname = f"{recid}_{start}_{end}.att.npy"
-                    att_path = os.path.join(args.out_dir, att_fname)
-                    model.save_attention_weight(att_path)
+        result = model.inference(
```
Comment (francoishernandez) [inline eend/chainer_backend/infer.py:107]: Not sure what we're doing here. Removing the whole chunking mechanism because it's supposed to be done before hand?

File: eend/chainer_backend/models.py [hunk 6]
```diff
@@ -71,6 +74,74 @@ def batch_pit_loss(ys, ts, label_delay=0):
     return loss, labels
 
 
+def _pit_loss_with_perm(pred, label, label_delay=0):
```
Comment (francoishernandez) [inline eend/chainer_backend/models.py:77]: Wasn't this loss already implemented? How was it done if not?

File: eend/chainer_backend/infer.py (hunk 5 above)
Comment (francoishernandez) [reply eend/chainer_backend/infer.py:107]: Ok, seen in models.py

File: eend/chainer_backend/models.py [hunk 7]
```diff
@@ -339,7 +410,71 @@ def dc_loss(embedding, label):
...
+                    shuffle=shuffle
+                )
+                if gpu >= 0:
+                    ys[0].to_cpu()
+                out_chunks.append(ys[0].data)
+                if save_attention_weight == 1:
+                    att_fname = f"{recid}_{start}_{end}.att.npy"
+                    att_path = os.path.join(out_dir, att_fname)
+                    self.save_attention_weight(att_path)
+        if hasattr(self, 'label_delay'):
```
Comment (francoishernandez) [inline eend/chainer_backend/models.py:461]: What's this "label delay" ?

File: eend/chainer_backend/models.py [hunk 8]
```diff
@@ -406,7 +541,302 @@ def save_attention_weight(self, ofile, batch_index=0):
...
+        with self.init_scope():
+            if initialW is None:
+                initialW = chainer.initializers.GlorotNormal()
+            self.emb = L.EmbedID(in_size, out_size, initialW=initialW)
+            self.alpha = chainer.variable.Parameter(
+                chainer.initializers.One(), (1, 1)
+            )
+            self.beta = chainer.variable.Parameter(
+                chainer.initializers.Zero(), (1, 1)
+            )
```
Comment (francoishernandez) [inline eend/chainer_backend/models.py:558]: Specific reason for one/zero ?

Comment (francoishernandez) [reply eend/chainer_backend/models.py:558]: By the way, these parameters are learned instead of tuned? Wouldn't this hurt what we "truly" want to learn -- the embedding?

File: eend/chainer_backend/models.py [hunk 9]
```diff
@@ -406,7 +541,302 @@ def save_attention_weight(self, ofile, batch_index=0):
...
+            if use_layer_norm:
+                self.layer_norm = L.LayerNormalization(out_size)
+        self._M = in_size
+        self.use_layer_norm = use_layer_norm
+
+    def forward(self, spk_embs, spk_ids):
+        """Compute distance between x and all word embedding in emb.
```
Comment (francoishernandez) [inline eend/chainer_backend/models.py:565]: word -> speaker ? :)

File: eend/chainer_backend/models.py [hunk 10]
```diff
@@ -406,7 +541,302 @@ def save_attention_weight(self, ofile, batch_index=0):
...
+            # distances = self.alpha * F.batch_l2_norm_squared(cur_emb - spk_emb) + self.beta
+
+            # BATCH
+            n_spks = len(spk_id)
+            # spk_all_emb: (M, E) -> (1, M, E) -> (C, M, E)
+            spk_all_emb = F.repeat(F.expand_dims(all_emb, axis=0), n_spks, axis=0)
+            # _spk_emb: (C, 1, E)
+            _spk_emb = F.expand_dims(spk_emb, axis=1)
+            # distances: norm((C, M, E) - (C, 1, E)) -> (C, M, E) -> (C*M, E) -> (C, M)
+            distances = spk_all_emb - _spk_emb
```
Comment (francoishernandez) [inline eend/chainer_backend/models.py:600]: Is this ok?

File: eend/chainer_backend/models.py [hunk 11]
```diff
@@ -406,7 +541,302 @@ def save_attention_weight(self, ofile, batch_index=0):
...
+                )
+                if gpu >= 0:
+                    ys[0].to_cpu()
+                    spk_emb[0].to_cpu()
+                out_chunks.append(ys[0].data)
+                out_spk_embs.append(spk_emb[0].data)
+                if save_attention_weight == 1:
+                    att_fname = f"{recid}_{start}_{end}.att.npy"
+                    att_path = os.path.join(out_dir, att_fname)
+                    self.save_attention_weight(att_path)
```
Comment (francoishernandez) [inline eend/chainer_backend/models.py:777]: Isn't this implemented in the main class? What are the differences that make it necessary to be implemented here?

**Review by francoishernandez — COMMENTED** [review]
First pass. Mostly questions. Probably missed a lot of things.

File: eend/feature.py (hunk 1 above)
Comment (Zenglinxiao) [reply eend/feature.py:277]: In the [EEND-vector paper](https://arxiv.org/abs/2010.13366), they mentioned at the end of section 2.3.1 diarization loss:

> If the number of speakers speaking in the chunk is smaller than S Local , we fill Y i with diarization label(s) of a virtual (S total + 1)-th always-silent speaker, i.e., (y t,i,S total +1 ∈ {0} | t = 1, . . . , T ).

To do that, I add a silent speaker as the first element of `kaldi_obj.global_speakers_with_silent`. So that, if we initialize `global_speaker_idx` with `np.zeros`, we make sure if the number of active speakers in a segment is smaller that set, those exceed will correspond to the silent speaker.

File: eend/feature.py (hunk 2 above)
Comment (Zenglinxiao) [reply eend/feature.py:281]: Yes. There may have some records that have actually more speakers than set. For example, we may have some record with 3 speakers but training on a model for 2 speakers.

File: eend/feature.py (hunk 4 above)
Comment (Zenglinxiao) [reply eend/feature.py:286]: `speaker_index` here is the (record) local speaker index, this is used to tag the reference.
For example, if we have found [A, C, D] speakers in record 0. While there are 26 speakers in total in the training set.
If for a segment/utterance 0 in record 0, there is [C, D] speaking, then:
* `speaker_index` will be [1, 2]. (A-0, C-1, D-2)
*  `global_speaker_index` will be [3, 4]. (silence-0, A-1, B-2, C-3,...)

File: eend/feature.py (hunk 3 above)
Comment (Zenglinxiao) [reply eend/feature.py:282]: This creates a map between `global_speaker_id` and local `speaker_id`.
Given a record with speaker [E, C, D].
`global_speaker_idx` here should give us [5, 3, 4] (silence-0, A-1, B-2, C-3,...)
Then the first speaker in the record is the 5th global speaker found in all training data. Which enables us to make it align when updating global_embeddings.

File: eend/chainer_backend/models.py (hunk 6 above)
Comment (Zenglinxiao) [reply eend/chainer_backend/models.py:77]: Yes. I just added a third return for the permutation corresponds to the lowest loss.
This info is needed later for the mapping between global speaker id and local speaker id.

File: eend/chainer_backend/models.py (hunk 7 above)
Comment (Zenglinxiao) [reply eend/chainer_backend/models.py:461]: This is actually never used throughout the project. I just leave it as is.

File: eend/chainer_backend/models.py (hunk 8 above)
Comment (Zenglinxiao) [reply eend/chainer_backend/models.py:558]: In the [EEND-vector paper](https://arxiv.org/abs/2010.13366), they mentioned in the last part of 2.3.2.Speaker embedding loss :

> Eq. (8) is the squared Euclidean distance between the learnable global speaker embedding and the estimated speaker embedding, which is rescaled with learnable scalar parameters α > 0 and β.

File: eend/chainer_backend/models.py (hunk 9 above)
Comment (Zenglinxiao) [reply eend/chainer_backend/models.py:565]: Yes, you're right.

File: eend/chainer_backend/models.py (hunk 11 above)
Comment (Zenglinxiao) [reply eend/chainer_backend/models.py:777]: Because the inference process of EEND-vector is a little bit different from the original EEND, as we need to produce also speaker embeddings. And I don't want to change it directly on infer.py since this may confuse people to think EEND also needs some ingredient that is actually specific to EEND-vector.

File: eend/chainer_backend/models.py (hunk 10 above)
Comment (Zenglinxiao) [reply eend/chainer_backend/models.py:600]: Double checked this, I think this conform to paper.
Could you elaborate more on what you think?
